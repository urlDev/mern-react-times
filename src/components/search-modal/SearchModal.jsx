import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import {
  fetchChartData,
  fetchRating,
  fetchForex,
  clearSearchResults,
  openSearchModal,
} from '../../redux/actions/chart';

import { SearchModalContainer, ResultsContainer } from './SearchModal.styles';

const SearchModal = () => {
  const { searchResults, chartTimeFrame } = useSelector((chart) => chart.chart);
  const dispatch = useDispatch();

  return (
    <SearchModalContainer>
      {searchResults.length ? (
        searchResults.map((stock) => (
          <ResultsContainer
            to={`/details/${stock.symbol.toLowerCase()}`}
            onClick={() => {
              dispatch(fetchForex(stock.symbol));
              dispatch(fetchChartData(stock.symbol, chartTimeFrame));
              dispatch(fetchRating(stock.symbol));
              dispatch(clearSearchResults());
              dispatch(openSearchModal());
            }}
            key={stock.symbol}
          >
            <h1>{stock.symbol}</h1>
            <h1>{stock.exchangeShortName}</h1>
          </ResultsContainer>
        ))
      ) : (
        <div
          style={{
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            height: '100%',
            color: 'white',
          }}
        >
          <h1 style={{ margin: '0' }}>No results found</h1>
        </div>
      )}
    </SearchModalContainer>
  );
};

export default SearchModal;
