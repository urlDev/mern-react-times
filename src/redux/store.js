import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));