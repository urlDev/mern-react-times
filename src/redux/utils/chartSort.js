const chartSort = (prevState, newState) => {
    const tempState = [...prevState, [...newState]];

    let sortedState = [...tempState];

    sortedState = sortedState
        .sort((a, b) => a[1] - b[1])
        .map((stock) => stock.slice(0, 1));

    return sortedState;
};

export default chartSort;