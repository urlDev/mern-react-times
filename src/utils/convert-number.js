const convertNumber = (labelValue) => {
  // Nine Zeroes for Billions
  return Math.abs(Number(labelValue)) >= 1.0e9
    ? (Math.abs(Number(labelValue)) / 1.0e9).toFixed(2) + "B"
    : // Six Zeroes for Millions
    Math.abs(Number(labelValue)) >= 1.0e6
    ? (Math.abs(Number(labelValue)) / 1.0e6).toFixed(2) + "M"
    : // Three Zeroes for Thousands
    Math.abs(Number(labelValue)) >= 1.0e3
    ? (Math.abs(Number(labelValue)) / 1.0e3).toFixed(2) + "K"
    : Math.abs(Number(labelValue));
};

export default convertNumber;
