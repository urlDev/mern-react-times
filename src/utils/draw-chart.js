// Reusable function to draw historical price chart in different components
const drawChart = ({ width, reDraw, chartData, svg, d3, index }) => {
  // reDraw => Charts in market details
  // because they are drawn according to screen sizes

  let height;
  let padding;
  let chartWidth;
  let dates;

  // reDraw is for when marketDetails page is shown
  // Chart is constantly redrawn then so thats why I made a separation
  if (reDraw) {
    // changing chart width according to screen width
    chartWidth = width < 1400 ? (width < 700 ? width - 30 : width - 200) : 550;
    height = width < 1400 ? chartWidth * 0.65 : 345;
    padding = 50;
  } else {
    // when its used in market components, width and height
    chartWidth = 100;
    height = 70;
    padding = 10;
  }

  svg.attr("width", chartWidth).attr("height", height);

  if (reDraw) {
    dates = chartData.map((value) => new Date(value.date));
  } else {
    // In market components, there are more than one chart
    // So I get their indexes from the component
    dates = chartData[index].map((value) => new Date(value.date));
  }

  // xScale
  const xAxisScale = d3
    .scaleTime()
    .domain([d3.min(dates), d3.max(dates)])
    .range(
      // Depending if its reDrawn or width changes,
      // In smaller screens, I don't put any padding
      reDraw
        ? width < 500
          ? [0, chartWidth]
          : [padding, chartWidth - padding]
        : [padding, chartWidth - padding]
    );

  // yScale
  const yAxisScale = d3
    .scaleLinear()
    .domain([
      // Chart is drawn from close data on each dataset
      // So in here I am checking where the function is being used
      d3.min(reDraw ? chartData : chartData[index], (data) => data.close),
      d3.max(reDraw ? chartData : chartData[index], (data) => data.close),
    ])
    .range([height - padding, padding]);

  // Charts in market cards don't use axes
  if (reDraw) {
    //   create x Axes
    const xAxes = d3
      .axisBottom(xAxisScale)
      .ticks(width < 700 ? d3.timeDay.every(10) : 5)
      .tickFormat((date) => d3.timeFormat("%b %d, %I:%M")(date));

    // draw x axes
    width > 500 &&
      svg
        .append("g")
        .call(xAxes)
        .attr("transform", `translate(0,${height - padding})`);

    // create y axes
    const yAxes = d3.axisLeft(yAxisScale).tickFormat((d) => `$ ${d}`);

    // draw y axes
    width > 500 &&
      svg.append("g").call(yAxes).attr("transform", `translate(${padding},0)`);
  }

  // create line
  const line = d3
    .line()
    .x((d) => xAxisScale(new Date(d.date)))
    .y((d) => yAxisScale(d.close));

  // draw historical price path
  const path = svg
    .append("path")
    .data(reDraw ? [chartData] : [chartData[index]])
    .style("fill", " none")
    .attr("stroke", reDraw ? "black" : "white")
    .attr("stroke-width", reDraw ? "1.5" : "1.2")
    .attr("d", line);

  // get the length of the path for line transition
  const pathLength = path.node().getTotalLength();

  // transition for the line
  const transitionPath = d3.transition().ease(d3.easeSin).duration(2000);

  // setting pathsLength for strokes attrs and using transition
  // I made it minus pathLength because it was drawing it backwards
  path
    .attr("stroke-dashoffset", -pathLength)
    .attr("stroke-dasharray", pathLength)
    .transition(transitionPath)
    .attr("stroke-dashoffset", 0);

  path.interrupt("stroke-dashoffset");
  path.interrupt("stroke-dasharray");
};

export default drawChart;
