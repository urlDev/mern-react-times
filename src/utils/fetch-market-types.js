import { changeMarketType, changeMarketName } from "../redux/actions/chart";

const fetchMarketTypes = (dispatch, market) => {
  //  One api endpoint can be used for all market quotes
  //  So I first determine the market type
  //  Then, checking which market type client has clicked on
  //  According to market, I pre-determined tickers/symbols
  //  With a switch function, I dispatch the related market types
  //  desired tickers to redux store.
  //  So with all this, I can use one fetch function for all market kinds,
  //  Therefore, less fetch calls, less code.
  //  More comments as it seemsüòÅ
  const Indexes = "%5EGSPC,%5ERUA,%5EDJI,%5ENDX,%5EN225,%5EFTSE";
  const Crypto = "BTCUSD,LTCUSD,XLMUSD,BCNUSD,ETHUSD,ETCUSD";
  const Forex = "EURUSD,USDJPY,GBPUSD,EURGBP,EURJPY,GBPJPY";
  const Stocks = "AAPL,FB,GOOG,TSLA,NFLX,AMZN";
  const Commodities = "ZGUSD,CLUSD,HGUSD,SIUSD,PLUSD,BZUSD";

  switch (market) {
    case "Indexes":
      return [
        dispatch(changeMarketType(Indexes)),
        dispatch(changeMarketName("Indexes")),
      ];
    case "Crypto":
      return [
        dispatch(changeMarketType(Crypto)),
        dispatch(changeMarketName("Crypto")),
      ];
    case "Forex":
      return [
        dispatch(changeMarketType(Forex)),
        dispatch(changeMarketName("Forex")),
      ];
    case "Stocks":
      return [
        dispatch(changeMarketType(Stocks)),
        dispatch(changeMarketName("Stocks")),
      ];
    case "Commodities":
      return [
        dispatch(changeMarketType(Commodities)),
        dispatch(changeMarketName("Commodities")),
      ];
    default:
      return Indexes;
  }
};

export default fetchMarketTypes;
