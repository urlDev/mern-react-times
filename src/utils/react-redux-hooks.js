// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
