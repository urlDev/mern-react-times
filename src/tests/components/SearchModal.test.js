import { mount, shallow } from "enzyme";
import { BrowserRouter as Router } from "react-router-dom";
import moxios from "moxios";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";
import {
  forexLengthOne,
  forexLengthTwo,
  rating,
  searchResult,
  searchResults,
} from "tests/fixtures/chart";

import SearchModal from "components/search-modal/SearchModal";
import { SearchModalAllPage } from "components/search-modal/SearchModal.styles";
import {
  clearSearchResults,
  CLEAR_SEARCH_RESULTS,
  CLOSE_SEARCH_MODAL,
  fetchForex,
  fetchRating,
  FETCH_FOREX_SUCCESS,
  SET_RATING,
} from "redux/actions/chart";

describe("Testing SearchModal component", () => {
  let wrapper;
  let store;

  beforeEach(() => {
    store = mockStore({
      searchResults,
      open: true,
    });

    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    wrapper = mount(
      <Router>
        <SearchModal store={store} />
      </Router>
    );

    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  test("Should render the component", () => {
    expect(wrapper).toHaveLength(1);
  });

  test("Should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Should close SearchModal on click", () => {
    wrapper.find("div").first().simulate("click");

    const actions = store.getActions();

    expect(actions).toEqual([{ type: CLOSE_SEARCH_MODAL }]);
  });

  test("Should dispatch actions when clicked on search result", async () => {
    wrapper.find("a").first().simulate("click");

    // When clicking on search result
    // It will dispatch two async functions
    // I return their results here
    // with mocking the process
    moxios.wait(() => {
      // at(0) and at(0) didnt work together
      // so I put mostRecent() for one of them
      const firstRequest = moxios.requests.mostRecent();
      firstRequest.respondWith({
        status: 200,
        response: forexLengthTwo.data,
      });
      const secondRequest = moxios.requests.at(1);
      secondRequest.respondWith({
        status: 200,
        response: rating,
      });
    });

    const expectedResults = [
      { type: CLEAR_SEARCH_RESULTS },
      { type: CLOSE_SEARCH_MODAL },
      { type: CLOSE_SEARCH_MODAL },
      { type: FETCH_FOREX_SUCCESS, payload: forexLengthTwo.data },
      {
        type: SET_RATING,
        payload: rating,
      },
    ];

    // async/await works just like return too
    await store.dispatch(fetchForex("TSLA"));

    const actions = store.getActions();
    expect(actions).toEqual(expectedResults);
  });
});
