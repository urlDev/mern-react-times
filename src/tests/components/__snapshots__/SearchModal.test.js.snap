// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchModal component Should render no results found if there are no results 1`] = `
<SearchModal
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <styled.div
    onClick={[Function]}
  >
    <div
      className="sc-dlfnbm iVzSdZ"
      onClick={[Function]}
    >
      <styled.div>
        <div
          className="sc-bdfBwQ kanTOC"
        >
          <div
            style={
              Object {
                "alignItems": "center",
                "color": "white",
                "display": "flex",
                "height": "100%",
                "justifyContent": "center",
              }
            }
          >
            <h1
              style={
                Object {
                  "margin": "0",
                }
              }
            >
              No results found
            </h1>
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</SearchModal>
`;

exports[`Testing SearchModal component Testing component with search results Should match snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <SearchModal
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <styled.div
        onClick={[Function]}
      >
        <div
          className="sc-dlfnbm iVzSdZ"
          onClick={[Function]}
        >
          <styled.div>
            <div
              className="sc-bdfBwQ kanTOC"
            >
              <Styled(Link)
                key="TSLA"
                onClick={[Function]}
                to="/details/tsla"
              >
                <Link
                  className="sc-gsTCUz gKMLZQ"
                  onClick={[Function]}
                  to="/details/tsla"
                >
                  <LinkAnchor
                    className="sc-gsTCUz gKMLZQ"
                    href="/details/tsla"
                    navigate={[Function]}
                    onClick={[Function]}
                  >
                    <a
                      className="sc-gsTCUz gKMLZQ"
                      href="/details/tsla"
                      onClick={[Function]}
                    >
                      <h1>
                        TSLA
                      </h1>
                      <h1>
                        NASDAQ
                      </h1>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
            </div>
          </styled.div>
        </div>
      </styled.div>
    </SearchModal>
  </Router>
</BrowserRouter>
`;
